physics/continuous_trajectory_body.hpp

```diff
@@ -442,32 +446,66 @@ ContinuousTrajectory<Frame>::ReadFromMessage(
             Polynomial<Position<Frame>, Instant>::template ReadFromMessage<
                 EstrinWithoutFMA>(polynomial));
       } else {
-        serialization::Polynomial const& polynomial = pair.polynomial();
-        if (polynomial.HasExtension(
-                serialization::PolynomialInMonomialBasis::extension) &&
+        not_null<serialization::Polynomial const*> polynomial =
+            &pair.polynomial();
+        bool const is_pre_καραθεοδωρή =
+            !polynomial->HasExtension(
+                serialization::PolynomialInMonomialBasis::extension) ||
             polynomial
-                .GetExtension(
+                ->GetExtension(
                     serialization::PolynomialInMonomialBasis::extension)
-                .has_evaluator()) {
-          // The post-Καραθεοδωρή path, do not specify an evaluator when calling
-          // `ReadFromMessage`.
+                .has_evaluator();
+        if (is_pre_καραθεοδωρή) {
           continuous_trajectory->polynomials_.emplace_back(
               Instant::ReadFromMessage(pair.t_max()),
-              Polynomial<Position<Frame>, Instant>::ReadFromMessage(
-                  polynomial));
+              UseHardwareFMA ? Polynomial<Position<Frame>, Instant>::
+                                   template ReadFromMessage<Estrin>(*polynomial)
+                             : Polynomial<Position<Frame>, Instant>::
+                                   template ReadFromMessage<EstrinWithoutFMA>(
+                                       *polynomial));
         } else {
-          // The pre-Καραθεοδωρή path.
+          std::optional<serialization::Polynomial> rewritten_polynomial;
```

Comment:
Let's not try to be smart in saving proto assignments.  Just do:
```
auto rewritten_polynomial = polynomial;
```
update it as needed below and then at line 501 just use `rewritten_polynomial`.